File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.cc
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(nullptr);

  int T;
  cin >> T;

  vector<int> counts;

  while (T--) {
    string W;
    int K;
    cin >> W >> K;

    int strlen = W.length();

    counts.assign(26, 0);
    for (const char &ch : W) {
      counts[ch - 'a'] += 1;
    }

    int minlen = 10001, maxlen = 0;
    for (int l = 0; l < strlen; l++) {
      if (counts[W[l] - 'a'] < K) {
        continue;
      }

      int sublen = 0;
      for (int r = l; r < strlen; r++) {
        if (W[l] == W[r]) {
          sublen += 1;
        }

        if (sublen == K) {
          minlen = min(minlen, r - l + 1);
          maxlen = max(maxlen, r - l + 1);
          break;
        }
      }
    }

    if (minlen == 10001 || maxlen == 0) {
      cout << -1 << "\n";
    } else {
    cout << minlen << " " << maxlen << "\n";
    }
  }

  return 0;
}

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.py
from collections import defaultdict

input = open(0).readline

T: int = int(input())

for _ in range(T):
    W: str = input()
    K: int = int(input())
    
    alpha_indices = defaultdict(list)
    for i, ch in enumerate(W):
        alpha_indices[ch].append(i)
    
    minlen: int = 10001
    maxlen: int = 0
    for indices in alpha_indices.values():
        if len(indices) < K:
            continue
        
        l: int = 0
        r: int = K - 1
        while r < len(indices):
            strlen: int = indices[r] - indices[l] + 1
            minlen = min(minlen, strlen)
            maxlen = max(maxlen, strlen)
            
            l += 1
            r += 1
    
    print(-1 if minlen == 10001 else f"{minlen} {maxlen}")